src/kagglehub/gcs_upload.py

```diff
@@ -72,18 +101,33 @@ def _upload_blob(file_path: str, model_type: str):
         token_exception = "'token' field missing from response"
         raise BackendError(token_exception)
 
-    headers = {"Content-Type": "application/octet-stream"}
+    session_uri = response["createUrl"]
+    headers = {"Content-Type": "application/octet-stream", "Content-Range": f"bytes 0-{file_size - 1}/{file_size}"}
+
+    retry_count = 0
+    uploaded_bytes = 0
 
-    # TODO(312511716): add resumable upload
-    with open(file_path, "rb") as f:
-        with tqdm(total=file_size, desc="Uploading", unit="B", unit_scale=True, unit_divisor=1024) as pbar:
-            reader_wrapper = CallbackIOWrapper(pbar.update, f, "read")
-            gcs_response = requests.put(response["createUrl"], data=reader_wrapper, headers=headers, timeout=600)
-            if gcs_response.status_code not in [200, 201]:
-                upload_fail_message = (
-                    f"Upload failed with status code: {gcs_response.status_code}, Response: {gcs_response.text}"
+    with open(file_path, "rb") as f, tqdm(total=file_size, desc="Uploading", unit="B", unit_scale=True) as pbar:
+        while uploaded_bytes < file_size and retry_count < MAX_RETRIES:
+            try:
+                f.seek(uploaded_bytes)
+                reader_wrapper = CallbackIOWrapper(pbar.update, f, "read")
+                headers["Content-Range"] = f"bytes {uploaded_bytes}-{file_size - 1}/{file_size}"
+                upload_response = requests.put(
+                    session_uri, headers=headers, data=reader_wrapper, timeout=REQUEST_TIMEOUT
                 )
-                raise BackendError(upload_fail_message)
+
+                if upload_response.status_code in [200, 201]:
+                    return response["token"]
+                elif upload_response.status_code == 308:  # Resume Incomplete # noqa: PLR2004
+                    uploaded_bytes = _check_uploaded_size(session_uri, file_size)
+                else:
+                    upload_failed_exception = f"Upload failed: {upload_response.text}"
+                    raise Exception(upload_failed_exception)
```

Comment:
Let's raise a BackendError instead and include the status code like before: https://github.com/Kaggle/kagglehub/blob/5df7847bb21116b7cc29a04af4d556ac8308716a/src/kagglehub/gcs_upload.py#L70